test/Npgsql.Tests/ConnectionTests.cs

```diff
@@ -24,35 +24,93 @@ public async Task BasicLifecycle()
         {
             using var conn = new NpgsqlConnection(ConnectionString);
 
-            bool eventOpen = false, eventClosed = false;
+            var eventOpen = false;
+            var eventClosed = false;
+
             conn.StateChange += (s, e) =>
             {
-                if (e.OriginalState == ConnectionState.Closed && e.CurrentState == ConnectionState.Open)
+                if (e.OriginalState == ConnectionState.Closed &&
+                    e.CurrentState == ConnectionState.Open)
                     eventOpen = true;
-                if (e.OriginalState == ConnectionState.Open && e.CurrentState == ConnectionState.Closed)
+
+                if (e.OriginalState == ConnectionState.Open &&
+                    e.CurrentState == ConnectionState.Closed)
                     eventClosed = true;
             };
 
             Assert.That(conn.State, Is.EqualTo(ConnectionState.Closed));
             Assert.That(conn.FullState, Is.EqualTo(ConnectionState.Closed));
 
-            conn.Open();
+            await conn.OpenAsync();
+
             Assert.That(conn.State, Is.EqualTo(ConnectionState.Open));
             Assert.That(conn.FullState, Is.EqualTo(ConnectionState.Open));
             Assert.That(eventOpen, Is.True);
 
-            using (var cmd = new NpgsqlCommand("SELECT 1", conn))
-            using (var reader = await cmd.ExecuteReaderAsync())
+            await using (var cmd = new NpgsqlCommand("SELECT 1", conn))
+            await using (var reader = await cmd.ExecuteReaderAsync())
             {
-                reader.Read();
+                await reader.ReadAsync();
+
                 Assert.That(conn.FullState, Is.EqualTo(ConnectionState.Open | ConnectionState.Fetching));
                 Assert.That(conn.State, Is.EqualTo(ConnectionState.Open));
             }
 
             Assert.That(conn.FullState, Is.EqualTo(ConnectionState.Open));
             Assert.That(conn.State, Is.EqualTo(ConnectionState.Open));
 
-            conn.Close();
+            await conn.CloseAsync();
+
+            Assert.That(conn.State, Is.EqualTo(ConnectionState.Closed));
+            Assert.That(conn.FullState, Is.EqualTo(ConnectionState.Closed));
+            Assert.That(eventClosed, Is.True);
+        }
+
+        [Test, Description("Makes sure the connection goes through the proper state lifecycle")]
+        //[Timeout(5000)]
+        public async Task BrokenLifecycle()
+        {
+            if (IsMultiplexing)
```

Comment:
I think this test should pass with multiplexing too, no? You probably need to make sure the connection is bound before getting its PID, that can be done simply by starting a transaction.

Reply:
You made me to find another bug :(